lib/stinger_core/inc/stinger_batch_insert.h

```diff
@@ -0,0 +1,571 @@
+                    assert(0);
+
+            }
+        }
+    }
+
+    static void
+    remap_results(iterator begin, iterator end)
+    {
+        OMP("omp parallel for")
+        for (iterator u = begin; u < end; ++u) {
+            int64_t result = adapter::get_result(*u);
+            switch (result){
+                case EDGE_ADDED:
+                case EDGE_UPDATED:
+                case EDGE_NOT_ADDED:
+                    // Subtract 10 to get back to return code caller expects
+                    adapter::set_result(*u, result - 10);
+                    break;
+                case -1:
+                case 1:
+                    // Caller must have set the return code, leave it be
+                    break;
+                case PENDING:
+                default:
+                    // We didn't finish all the updates, or invalid codes were set
+                    assert(0);
+
+            }
+        }
+    }
+
+    static void
+    batch_update_dispatch(stinger_t * G, iterator updates_begin, iterator updates_end, int64_t operation, bool directed)
+    {
+        // Each edge is stored in two places: the out-edge at the source, and the in-edge at the destination
+        // First iteration: group by source and update out-edges
+        // Second iteration: group by destination and update in-edges
+        // But, in order to avoid deadlock with concurrent deletions, we must always lock edges in the same order
+        LOG_V("Partitioning batch to obey ordering constraints...");
+        const iterator pos = std::partition(updates_begin, updates_end, source_less_than_destination);
+
+        const int64_t OUT = STINGER_EDGE_DIRECTION_OUT;
+        const int64_t IN = STINGER_EDGE_DIRECTION_IN;
+
+        // All elements between begin and pos have src < dest. Update the out-edge slot first
+        LOG_V("Beginning OUT updates for first half of batch...");
+        do_batch_update<OUT, source_funcs, dest_funcs>(G, updates_begin, pos, operation);
+        clear_results(updates_begin, pos);
+        LOG_V("Beginning IN updates for first half of batch...");
+        do_batch_update<IN, dest_funcs, source_funcs>(G, updates_begin, pos, operation);
+
+        // All elements between pos and end have src > dest. Update the in-edge slot first
+        LOG_V("Beginning IN updates for second half of batch...");
+        do_batch_update<IN, dest_funcs, source_funcs>(G, pos, updates_end, operation);
+        clear_results(pos, updates_end);
+        LOG_V("Beginning OUT updates for second half of batch...");
+        do_batch_update<OUT, source_funcs, dest_funcs>(G, pos, updates_end, operation);
+
+        // For undirected, do the same updates again in the opposite direction
+        if (!directed)
+        {
+            // TODO does anyone really care what happens when half of an undirected insert has a different return code?
+            clear_results(updates_begin, updates_end);
+
```

Comment:
I'm ok with leaving this as is. Maybe we make a note somewhere that officially a STINGER graph is to consist entirely of either directed or undirected edges, not a combination.